### PR TITLE
User/UserService

### PR DESCRIPTION
**Pull request by msladek**
https://synjira.atlassian.net/browse/CELDEV-690
Introduces the `UserService` and `User` object. It ended up causing many refactorings in different commands to consolidate behaviour in the new service. sorry about the huge PR.

### CODE REVIEW

File: src/main/java/com/celements/auth/user/CelementsUserService.java [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.objects.BaseObject;
+import com.xpn.xwiki.user.api.XWikiGroupService;
+import com.xpn.xwiki.user.api.XWikiUser;
+import com.xpn.xwiki.web.Utils;
+
+@Component
+public class CelementsUserService implements UserService {
+
+  private static Logger LOGGER = LoggerFactory.getLogger(CelementsUserService.class);
```
Comment (fpichler) [inline src/main/java/com/celements/auth/user/CelementsUserService.java]: Please add final

File: src/main/java/com/celements/auth/user/CelementsUserService.java [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+
+  @Override
+  public User createNewUser(String accountName, Map<String, String> userData, boolean validate)
+      throws UserCreateException {
+    userData.put("xwikiname", checkNotNull(Strings.emptyToNull(accountName)));
+    return createNewUser(userData, validate);
+  }
+
+  @Override
+  public synchronized User createNewUser(Map<String, String> userData, boolean validate)
```
Comment (fpichler) [inline src/main/java/com/celements/auth/user/CelementsUserService.java]: Why synchronized?

File: src/main/java/com/celements/auth/user/CelementsUserService.java [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+    }
+  }
+
+  DocumentReference getOrGenerateUserDocRef(String accountName) {
+    accountName = Strings.nullToEmpty(accountName);
+    if (accountName.isEmpty()) {
+      accountName = RandomStringUtils.randomAlphanumeric(12);
+    }
+    DocumentReference userDocRef = resolveUserDocRef(accountName);
+    while (modelAccess.exists(userDocRef)) {
```
Comment (fpichler) [inline src/main/java/com/celements/auth/user/CelementsUserService.java:171]: Maybe we need to get a XWiki lock in the new docref to ensure mutex on clusters.

**Review by fpichler — CHANGES_REQUESTED** [review]

File: src/main/java/com/celements/auth/user/CelementsUserService.java (hunk 3 above)
Comment (msladek) [reply src/main/java/com/celements/auth/user/CelementsUserService.java:171]: this functionality seems similar to `NextFreeDocService`, which has the same issue. perhaps we should add a lock there and provide a method for random alphanumeric names to be used here? wdyt?

Comment (fpichler) [reply src/main/java/com/celements/auth/user/CelementsUserService.java:171]: ok, I agree. Yet, it is not part of the Issue adressed in this PR. Thus please add a TODO for a corresponding new Issue and I will approve the PR.

File: src/main/java/com/celements/auth/AuthenticationService.java [hunk 4]
```diff
@@ -1,45 +1,64 @@
...
-import com.xpn.xwiki.objects.BaseObject;
 import com.xpn.xwiki.objects.classes.PasswordClass;
 import com.xpn.xwiki.user.api.XWikiUser;
-import com.xpn.xwiki.user.impl.xwiki.XWikiRightServiceImpl;
 
 @Component
 public class AuthenticationService implements IAuthenticationServiceRole {
 
-  private static Logger _LOGGER = LoggerFactory.getLogger(AuthenticationService.class);
+  private static Logger LOGGER = LoggerFactory.getLogger(AuthenticationService.class);
```
Comment (fpichler) [inline src/main/java/com/celements/auth/AuthenticationService.java]: please add final

**Review by fpichler — CHANGES_REQUESTED** [review]

File: src/main/java/com/celements/auth/user/CelementsUserService.java (hunk 2 above)
Comment (msladek) [reply src/main/java/com/celements/auth/user/CelementsUserService.java]: to ensure at least a system wide mutex. but moved to `getOrGenerateUserDocRef`

File: src/main/java/com/celements/auth/user/CelementsUserService.java (hunk 3 above)
Comment (msladek) [reply src/main/java/com/celements/auth/user/CelementsUserService.java:171]: Issue created: https://synjira.atlassian.net/browse/CELDEV-692

File: src/main/java/com/celements/web/service/CelementsWebService.java [hunk 5]
```diff
@@ -14,81 +14,73 @@
...
-import com.xpn.xwiki.objects.BaseObject;
 import com.xpn.xwiki.user.api.XWikiUser;
 import com.xpn.xwiki.util.Util;
 import com.xpn.xwiki.web.XWikiResponse;
 
 @Component
 public class CelementsWebService implements ICelementsWebServiceRole {
 
-  private static Logger _LOGGER = LoggerFactory.getLogger(CelementsWebService.class);
+  private static Logger LOGGER = LoggerFactory.getLogger(CelementsWebService.class);
```
Comment (fpichler) [inline src/main/java/com/celements/web/service/CelementsWebService.java]: please add final

**Review by fpichler — CHANGES_REQUESTED** [review]

**Review by ebeutler — DISMISSED** [review]

**Review by fpichler — APPROVED** [review]

**Pull request merged**